week3/src/components/CommonPage.jsx

```diff
@@ -0,0 +1,105 @@
+import styled from "styled-components";
+import { useEffect, useState, useMemo } from "react";
+import SingleCard from "./SingleCard";
+import Header from "./Header";
+import Modal from "./Modal";
+import ModalPortal from "./ModalPortal";
+
+const CommonPage = (cardList) => {
+  const [choiceOne, setChoiceOne] = useState(null);
+  const [choiceTwo, setChoiceTwo] = useState(null);
+  const [counter, setCounter] = useState(0);
+  const [modalOn, setModalOn] = useState(false);
+  const [disabled, setDisabled] = useState(false);
+
+  const flippedCard = document.getElementsByClassName("flipped");
+
+  // useMemo() 활용하여 cardList가 변경되지 않을 경우, 이전 값을 재사용하도록 구현
+  const copiedCardList = useMemo(() => {
+    setCounter(0);
+    // JSON.parse(JSON.stringify(obj)): 깊은 복사
+    const copied = JSON.parse(
+      JSON.stringify(
+        // Object.keys(): 객체를 문자열 배열로 변환
+        Object.keys(cardList)
+          .map((item) => cardList[item])
+          // flat(): 하나의 배열로 만들고자 사용
+          .flat()
+      )
+    );
+    // 카드 선택을 하다가 중간에 reset 버튼이나 레벨 선택을 다시 하면 이전에 저장되어 있던 카드 선택 정보 모두 삭제
+    copied.map((it) => (it.matched = false));
+
+    return copied.sort(() => Math.random() - 0.5);
+  }, [cardList]);
+
+  // 값이 null이 아니면 이미 해당 값은 선택되어 있다는 것
+  const handleChoice = (card) => {
+    choiceOne ? setChoiceTwo(card) : setChoiceOne(card);
+  };
+
+  useEffect(() => {
+    if (choiceOne && choiceTwo) {
+      // 두 개의 카드가 같은지 확인하는 동안 다른 카드 선택 불가
+      setDisabled(true);
+
+      if (choiceOne.name === choiceTwo.name) {
+        setChoiceOne((choiceOne.matched = true));
+        setChoiceTwo((choiceTwo.matched = true));
+        setCounter((prev) => prev + 1);
+        resetTurn();
+
+        flippedCard.length === copiedCardList.length
+          ? setModalOn(true)
+          : setModalOn(false);
```

Comment:
```suggestion
    setModalOn(flippedCard.length === copiedCardList.length);
```
요렇게 하면 좀 더 간결하게 쓸 수 있겠당!ㅎㅎ